### PR TITLE
Fix Base.print_matrix() for big matrices

### PR DESCRIPTION
**Pull request by artkuo**
The REPL is slow displaying large vectors or matrices, e.g. `eye(10^4)`. The reason is that Base.print_matrix() has to figure out how to display only one screenful of properly aligned columns, but in doing so, it evaluates the spacing (alignment) of all the elements, even when there are 10^10 of them. The solution is to have Base.print_matrix() evaluate a much smaller (though still conservative) subset of the matrix, so that sufficient alignment info is determined, but not overly much.

Base.print_matrix() was not originally tested, so new tests are added, including matrices that fit on the screen, are too high, too wide, or both.

This problem was discovered while developing Base.print_range() #13534, and the fix has been brought back to print_matrix.

### CODE REVIEW

File: base/show.jl [hunk 1]
```diff
@@ -1062,79 +1062,98 @@ function print_matrix_vdots(io::IO,
...
                 end
             end
-        else # neither rows nor cols fit
-            c = div(cols-length(hdots)+1,2)+1
-            R = reverse(alignment(X,I,n:-1:1,c,c,ss))
-            c = cols - sum(map(sum,R)) - (length(R)-1)*ss - length(hdots)
-            L = alignment(X,I,1:n,c,c,ss)
-            r = mod((length(R)-n+1),vmod)
-            for i in I
+        else # neither rows nor cols fit, so use all 3 kinds of ellipsis
```
Comment (stevengj) [inline base/show.jl]: The plural of "ellipsis" is "ellipses".

**Comment by artkuo** [comment]
No idea why this failed automatic check. I was able to make and test on my machine, and everything passed with no problems.

File: base/show.jl (hunk 1 above)
Comment (artkuo) [inline base/show.jl]: Yeah, but I like "ellipses" to mean the plural of those oval thingees. It hurts my head to use it to mean plural of ellipsis. Next edit I will change this to dots. But meanwhile there are bigger problems, as in despite this building successfully on my machine, it seems to fail automatic checks, for reasons beyond me. If you have any advice on how to figure this out, I'd much appreciate it.

**Comment by stevengj** [comment]
@artkuo, the failure is #11553, not your fault.

**Comment by jakebolewski** [comment]
@artkuo this is a great improvement, thanks! (feel free to clean up more display code :wink:)

**Comment by KristofferC** [comment]
Performance increase, clearer code, better documentation, tests added... Nice job!

**Pull request merged**